### ISSUE TITLE
The best way dealing with sparse Behavior/Decorator?

### ISSUE THREAD

**Issue by ThisNoName** [jbogard/MediatR#248]
I tried behavior and it works of course.  But say 5% of the requests require caching, then the other 95% of the traffic are forced to check CachingAttribute and quit.  If you have a bunch of such behaviors, there're lots of wasted effort.

In the good'n'old days, I used SimpleInjector to add conditional decorators during startup, so only requests marked with Caching are actually decorated.  But in this AspNetCore age, Microsoft's IoC cannot do conditional registration, and due to restriction in .NET, SimpleInjector cannot handle injection into MVC controller and attribute.   If I want the old way, I have to I code two separate sets of IoC for MVC and MediatR.  Doesn't feel right either.

Are there any more efficient approach?  For example, does MediatR have something similar to aspnetcore TypeFilterAttribute?  Basically, you throw a passive attribute and an implementation onto any requests, and it auto wires up and only run the code on decorated requests.  

I think aspnetcore does this with FilterFactory outside regular IoC.  I would imagine similar can be done using a generic behavior with reflection following some convention?  But between efficiency and all the complexity of async, I'm a little afraid of DIY this for PROD.  

Any thoughts?  Maybe this can be a useful new feature?  Or do I worry to much, just let the server run through a dozen behaviors on every request?

Thanks.

**Comment by ThisNoName** [comment]
Back in the days when I have 100% of my code in MediatR project, that would work well.  But some of them are just too simple to deserve living in another layer.  Now I only reserve MediatR as business layer for things that have at least some logic, and leave the rest of the simple CRUD (especially read) in MVC controller.  But aspnetcore put some serious limitation on what other IoC can do.

I'm a big fan of the TypeFilterAttribute approach, throw in a passive attribute and a implementation class, and all is taken care of.  I think it's just like behavior, except with auto and conditional registration.  Would love to see that in MediatR.

Either way, thanks for the great work over the years.

**Comment by ThisNoName** [comment]
Microsoft throw in [ServiceFilter and TypeFilter](https://docs.microsoft.com/en-us/aspnet/core/mvc/controllers/filters#dependency-injection) in the latest MVC for crosscutting concerns.  I like TypeFilter better because it's instantiated via DependencyInjection.ObjectFactory, so you don't even have to register, just throw it anywhere and it just works.  The other great thing is that you can fine-tune control with additional parameters.

A standard Microsoft implementation of TypeFilterAttribute looks like this, which many consider super ugly

```
[TypeFilter(typeof(LogFilter), Arguments = new object[] {"myLogger"})]  
public IEnumerable<Item> Get() { ... }
```

Of course the Internet always come up with something better.  I forgot who posted it first, my final implementation looks like this: 

```
public class DependencyAttribute : TypeFilterAttribute {

public DependencyAttribute(DependencyEnum dependency) : 
   base(typeof(DependencyAttributeImpl))  {
      base.Arguments = new object[] { dependency }; }

private class DependencyAttributeImpl : IAsyncActionFilter {
   public DependencyAttributeImpl(
      ISystemStatusService system, // Via standard .NET IoC
      DependencyEnum dependency // Optional parameter from attribute ) 
      {  ... }

   public async Task OnActionExecutionAsync(
          ActionExecutingContext context, ActionExecutionDelegate next)  {
      
        if (!_system.IsAvailable(_dependency, out DependencyEnum active)) {
         context.HttpContext.Response.StatusCode = 503;
         context.Result = "Say something";
         return;  }
      await next(); }}}
```

Now I can do this on my controller/action

```
[Dependency(DependencyEnum.SysA | DependencyEnum.SysB)]
public IEnumerable<Item> Get() { ... }
```

I like this approach because it's kind of keeping the spirit of passive attribute, but does not have to fire up every crosscutting code only to check there's no corresponding custom attribute and drop out. 

I can still do this with SimpleInjector and decorator like last time, but then I have to wire IoC twice because no one can touch controller if I need any injection there.  [https://github.com/jbogard/MediatR/issues/55](url)

I would imagine a CrossCuttingBehavior could do some auto wiring similar to IoC.ObjectFactory?  Or maybe something deeper in your framework that can handle this more efficiently.  

Basically, throw a bunch custom attributes on any request, and corresponding handlers/decorators fire up along with rest of dependencies, no hands ^_^

**Comment by ThisNoName** [comment]
Saw this is still open in the age of 5.0 ^_^

Basically, I'm looking for decorator with passive attribute.  This is what I did earlier using SimpleInjector.  

```
container.RegisterDecorator(typeof(IRequestHandler<,>), typeof(TransactionRequestHandler<,>),
   c => c.ImplementationType.GetInterfaces().Where(i => i.IsGenericType 
                && i.GetGenericTypeDefinition() == typeof(IRequestHandler<,>))
        .SelectMany(i => i.GetGenericArguments()).First()
        .UnderlyingSystemType.Attributes(typeof(TransactionAttribute)).Any());
```
It basically says if there's a TransactionAttribute on the request, then decorate corresponding handler.  So those without are never decorated, and don't have to bother checking the non-existing attribute then quit.

Now that M$ killed all IoC with their lowest common denominator injection, if I want to do the same, I have to register everything twice.  Not that it is hard, but maybe decorator is a common enough scenario to deserve its own place in the framework like the pipeline?

Feel free to point out the fault in my logic.  I'm really a novice programmer 20 years in the making ^_^

**Comment by jbogard** [comment]
Definitely go pipeline. I created that explicitly because decorators
weren’t flexible enough.

**Comment by ThisNoName** [comment]
Isn't every pipeline end up running for every request?  Say 5% of my requests need Transaction, the other 95% also have to go through the pipeline only to see there's no TransactionAttribute.  And this would be true for all sparsely needed passive attributes, and that could be a lot of reflection for nothing.  Is performance a concern here?  Not that it actually matters in my small application, just curious.

I actually had an implementation that creates a dictionary of Request/TransactionAttribute at Startup, then in the pipeline, go by the dictionary instead of running reflection every time.  But I thought that was too hackish and wasn't sure if it's production worthy.

I recently learned that .NET IoC can inject IEnumerable<Something>, so one alternative might be a single pipeline, and map whatever attributes present to corresponding handling classes by convention.  But that might not be available to all IoC framework and injecting a long list of dependencies isn't exactly free.

Or would you suggest just wire up whatever pipelines needed and let Moor's Law does the rest ^_^

**Comment by jbogard** [comment]
So *today* MediatR still wholly relies on the container to determine what pipeline behaviors should be instantiated. I had considered making the pipeline more complicated - so that you explicitly configure through MediatR, instead of through the container, which pipeline steps should execute.

But then that got way more complicated so I gave up :)

If you find that expensive, perhaps you can cache that lookup in a `ConcurrentDictionary`? I'm not sure how much faster that would be, however.

Another option would be to use a marker interface instead of an attribute. Then you could use generic constraints. Most containers can handle that, only picking the pipeline behaviors that can fulfill the constraint (in fact, I submitted a PR to allow the MS ASP.NET Core DI container to handle it).
